Added TF parsing

ShobhitBehl opened the pull request:
Added parsing for TF files (SavedModel and FrozenGraph)

ShobhitBehl committed:
Add TFLite

Added code to parse a .tflite file into a Graph object,
extracting operator attributes onto nodes and extracting tensor
attrubtes onto edges. Added build rules for the same.

ShobhitBehl committed:
Bug fix and added init.py

Removed auto generated files, added a module(init.py) to generate
these files from schema.fbs using flatc compiler.
Added build rules for the same.

ShobhitBehl committed:
Update README.md

ShobhitBehl committed:
Bug fix and update .gitignore

Fixed error in TFLite/src/init.py, updated .gitignore, added more
comments and fixed import format across files.

ShobhitBehl committed:
Added common sub-directory and comments

Added a common sub-directory to store generic modules Graph, Node and
Edge.
Added links in comments in Node and OpToNode for operator support and
semantic meaning of attributes
Added function to calculate duplication of operators and tensors
accross all files in models.

ShobhitBehl committed:
Remove some attributes from Node

Remove unecessary attributes from Node, update .gitgnore.

ShobhitBehl committed:
Add Schema_DDL, Storage.py, argparse in main.py

Added Storage module which stores a graph into a spanner database, added the
same to BUILD file.
Added Schema_DDL used to create the DB tables.
Added argparse in main.py for passing filename, model_name and category as
arguments.
Updated attribute names to match with the database columns.

ShobhitBehl committed:
Added parsing of TF files

Added TF directory with modules for parsing saved model and frozen graph
files, added additional attributes in Node for the same.
Bug fixes in common/Graph.py and common/Storage.py
Changes in Schema_DDL in accordance to attributes added and removed in
Node.
Added exmaple queries for Spanner.

ShobhitBehl committed:
Restructure code

ShobhitBehl committed:
Add source field in common/Graph.py

Added a source attribute to denote parsed file format in
common/Graph.py.
Added a mapping from TFLite operator naming to TF counterparts to
maintain consistency in DB.

ShobhitBehl committed:
Added Exception handling in common/Storage.py

Added exception handling to maintain DB consistency in
common/Storage.py.

ShobhitBehl committed:
Added Models and updated common/visualize.ipynb

ShobhitBehl committed:
Added Docstrings

Added Docstrings to all code files.
Added some defaults for argparse.
Naming changes, code clean up.

rachitagrawal commented on TF/src/TFParser.py:
```diff
@@ -202,6 +203,15 @@ def parse_graph(self, file_path, model_name, category, is_saved_model, input_ope
+            #                 print(nodes[node_index].operator_type)
+            #                 if nodes[node_index].operator_type == "MatMul":
+            #                     print("Got One!")
+
```
May be remove this if this is not needed.

ShobhitBehl replied:
Fixed in latest commit

ShobhitBehl replied:
Already fixed in a previous commit

rachitagrawal commented on Queries.sql:
```diff
@@ -0,0 +1,83 @@
+  Operators.model_name = Tensors.model_name
+  AND operator_id IN UNNEST(from_operator_ids)
+WHERE
+  operator_type = "Input_Placeholder"
```
Some description about each query would be helpful.

ShobhitBehl replied:
Fixed in latest commit

ShobhitBehl replied:
Added in a previous commit

rachitagrawal commented on TF/src/common/Graph.py:
```diff
@@ -0,0 +1,137 @@
+        return ret    
+
+    # Print graph in BFS order
+    def print_graph(self):
```
This is pretty much same as in TFLite/src/common/Graph.py. How about having a common util where we can move such common utility and reduce code duplication?

ShobhitBehl replied:
Fixed in commit with message 'Restructure code', changed file structure for common files.

rachitagrawal commented on TF/src/common/Graph.py:
```diff
@@ -0,0 +1,137 @@
+                        visited[dest_node_index] = True
+                        queue.put(dest_node_index)
+        
+        return list(traversed_edges)
```
Same for other functions as well. See if we can reduce some code duplication. 'print_graph' and 'get_traversed_graph' is same and repeated for each TFLite and TF.

ShobhitBehl replied:
Will try to address 'print_graph' and 'get_traversed_graph' code duplication in the subsequent commits.

rachitagrawal commented on TF/src/common/Storage.py:
```diff
@@ -0,0 +1,208 @@
+                    )
+                    tensor_id += 1
+
+    # Function to load data for given Graph to Spanner Database
```
Why not pull all common stuff in to a base class and the derive it for both TF and TFLite. Again lot of code duplication.

ShobhitBehl replied:
Fixed in commit with message 'Restructure code', changed file structure for common files.

rachitagrawal commented on TF/src/main.py:
```diff
@@ -0,0 +1,50 @@
+    parser.add_argument('model_name')
+    parser.add_argument('category') 
+    parser.add_argument('is_saved_model')
+    args = parser.parse_args()
```
Add default values. Ignore if already added in the subsequent commit.

ShobhitBehl replied:
Added default values.

rachitagrawal commented on common/Storage.py:
```diff
@@ -61,7 +61,7 @@ def _load_operators(self, graph):
 
         # Number of nodes to be processed per batch 
-        num_attributes = len(vars(graph.nodes[0]))
+        num_attributes = len(vars(graph.nodes[0])) + 2
```
Added a comment about + 2.

ShobhitBehl replied:
Added comment

rachitagrawal commented on common/Storage.py:
```diff
@@ -152,7 +152,7 @@ def _load_tensors(self, graph):
 
         # Number of edges to be processed per batch
-        num_attributes = len(vars(graph.edges[0]))
+        num_attributes = len(vars(graph.edges[0])) + 4
```
Same here.

ShobhitBehl replied:
Added comment

rachitagrawal reviewed (commented):
Added some comments. Ignore the ones that you have already taken care of. Saw some are already covered in the last commit.

ShobhitBehl committed:
Add case handling for StatefulPartitionedCall

Added case handling for graphs with 'StatefulPartitionedCall' operation
as these graphs are not supported.
Fixed import order.
Changed variable naming from 'columns' to 'column_names' in
common/Storage.py.
Added model directory as a constant literal in main.py of TF and TFLite.
Added a corner case check in case of tf.get_static_value() returning
None.
Added support for operations FusedBatchNormV2 and FusedBatchNormV3.
Added more models in load_data.sh.

rachitagrawal reviewed (approved):
Do remember to 'rebase' it before merging after all the approvals.

prakashprabhu commented on Queries.sql:
```diff
@@ -0,0 +1,40 @@
+GROUP BY model_name
+ORDER BY num_conv2d_layers DESC;
+
+-- Operator Distribution per problem category
```
nit: These are for TFLite only? Perhaps add in comment?

ShobhitBehl replied:
Added comment

prakashprabhu commented on Queries.sql:
```diff
@@ -0,0 +1,40 @@
+GROUP BY model_name, activation_function
+ORDER BY model_name, count DESC;
+
+-- Padding types used per operator per model
```
Same as above.

ShobhitBehl replied:
Added comment

prakashprabhu commented on TF/src/OpToNode.py:
```diff
@@ -0,0 +1,102 @@
+        elif op == "Cast":
+            node.in_data_type = str(attr['SrcT'])[9:-1]
+            node.out_data_type = str(attr['DstT'])[9:-1]
+
```
nit: Do you want to emit a warning here in case you find an op that is not the list above?

ShobhitBehl replied:
Have not added a warning yet, will be printed a lot of unnecessary warnings for a given graph since many operations do not contain any extractable attributes. Will look to solve the same in subsequent commits.

ShobhitBehl committed:
Add comments for batch computation in Storage.py

Added comments for batch computation in common/Storage.py.
Added const list is common/Storage.py for attributes not being added to
database.
